### ISSUE TITLE
Support safe shift-left in solc 0.8.x

### ISSUE THREAD

**Issue by barakman** [ethereum/solidity#11447]
## Abstract

The shift-left operation can overflow.

Yet, unlike all other arithmetic operations of this nature, shift-left does not revert upon overflow.

## Motivation

Starting from solc 0.8.0, every arithmetic operation which has the potential to overflow, reverts upon overflow.

Every arithmetic operation - except for shift-left.

In other words, every expression consisting of `<<` or `<<=` may yield incorrect result.

[The official documentation](https://docs.soliditylang.org/en/v0.8.0/control-structures.html#checked-or-unchecked-arithmetic) is clear about this fact.

Nevertheless, anyone unfamiliar with it, is in for a world of hurt.

## Specification

My suggestion is to align the behavior of shift-left upon overflow to be compatible with that of all other arithmetic operations.

## Side Note

The official documentation also lists division and modulo as arithmetic operations which are checked starting from solc 0.8.0.

While this may be correct in the sense that these two operations will from now on return a 'revert' opcode instead of an 'invalid' opcode (which impacts the gas cost inflicted on the caller), the context of this entire section is overflow and underflow handling.

And under this context, it seems inappropriate to mention these two operations, since they obviously never overflow or underflow (at least when the operands are `unsigned`, I'm not really sure about the definition of 'overflow' and 'underflow' when the operands are `signed`).

So my additional suggestion is to somehow differentiate these operations from all the rest, explain what 'checked' means with respect to them, and perhaps even elaborate on the 'revert vs invalid' opcode issue.

**Comment by chriseth** [comment]
Can you provide some examples where the "overflow" check for shift would be useful? Since shifts operate on the bit-representation, we assumed that truncation is the expected behaviour. I currently cannot think of a situation where you would want to check for bits being shifted out of the range.

Also, just for clarification, I myself consider shift a bit operation instead of an arithmetic operation, since it is usually not defined based on its arithmetic/numeric semantics, but on the semantics of the bit representation of the values.

About your side note: This is solely for the case `type(int).min / -1`. I think for `%` overflow can actually never occur. Still it might be useful to list all the arithmetic operations.

**Comment by barakman** [comment]
Consider a function which computes the integer division of a 512-bit dividend by a 256-bit divisor.

If the result of the computation is larger than 256 bits, then we want this function to revert.

Here is the function, which implements a simple long-division method:
```
function div(uint256 xh, uint256 xl, uint256 y) returns (uint256) {
    uint256 result = 0;
    uint256 length = 255 - log2(y);
    while (xh > 0) {
        uint256 bits = log2(xh) + length;
        result += 1 << bits;
        (xh, xl) = sub(xh, xl, y >> (256 - bits), y << bits);
    }
    return result + xl / y;
}
```
Note that we have two shift-left operations in the function above:
- `1 << bits`
- `y << bits`

The purpose of `y << bits`, just like you mentioned, is to obtain the LSB part of `y`.

So we indeed want this expression to overflow **without** reverting (for which purpose, we can use `unchecked`).

In contrast, if `1 << bits` overflows, then it means that the result of the function is going to be larger than 256 bits.

Hence we want it this expression to revert on overflow.

Thanks

**Comment by hrkrshnn** [comment]
@barakman While this does not answer you question, if you need overflow checking for `1 << bits`, you can replace it by `2**y`. For the IR based codegen and yul optimizer (`--experimental-via-yul`), it would actually lead to efficient codegen. That is, change the exponentiation to `shl`, except that it reverts on overflow. Unfortunately, for the current code gen, the `exp` still remains (need to improve the inliner to fix this.)

On an unrelated note, can we talk about how you use 512 bit division? It would be great if you can ping me @hrkrshnn or @hrkrshnn:matrix.org on gitter / matrix.

**Comment by barakman** [comment]
@hrkrshnn:
My purpose in this feature request is not to seek a solution for asserting on shift-left, but rather to align the behavior of this arithmetic operation, with the behavior of all other arithmetic operations **at the compiler-level** (behavior under the context of overflow handling of course).

My concern is that for those who are oblivious of this inconsistency, thus assuming `<<` to revert on overflow.

With regards to your other question - I use this function in order to extend the input range in the expression `x * y / z` (I have left out the entire code, since it is not directly related to the issue at hand).

**Comment by hrkrshnn** [comment]
I agree that the behaviour of shifts doesn't fit with the rest. In fact, at least two other users have pointed this out recently. Perhaps we should indeed consider reverting shifts.

**Comment by barakman** [comment]
@hrkrshnn:
All of that being said, I do have to agree with the point made earlier by @chriseth, because the shift-left operation is often regarded as a bit-wise operation rather than an arithmetic operation, in the sense that it is not designated for a quantitive computation, but rather for retrieving a specific part in the 2s-complement representation of an integer, in which case, the term 'overflow' is de facto meaningless (apologies for the longest sentence in recorded history).

TBH, even in the example that I gave earlier in this thread, I have used shift-left in one case for a quantitive computation (overflow meaningful), and in the other case in order to obtain a part of the bit-representation of the input value (overflow meaningless).

The thing is - while the bit-representation can be computed by executing it inside an `unchecked` scope, the quantitive computation cannot be computed (safely) by executing it outside an `unchecked` scope, because the compiler does not revert on that.

So one needs to perform the assertion as part of the code, which at this point (solc 0.8.x) does not align with all other operations.
